timemachine/cpp/src/hilbert_curve.cu

```diff
@@ -0,0 +1,85 @@
+#include "gpu_utils.cuh"
+#include "hilbert_curve.hpp"
+#include "kernels/k_hilbert.cuh"
+#include "vendored/hilbert.h"
+#include <cub/cub.cuh>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+namespace timemachine {
+
+HilbertCurve::HilbertCurve(int N)
+    : N_(N), d_bin_to_idx_(HILBERT_GRID_DIM * HILBERT_GRID_DIM * HILBERT_GRID_DIM), d_sort_keys_in_(N),
+      d_sort_keys_out_(N), d_sort_vals_in_(N), d_sort_storage_(nullptr), d_sort_storage_bytes_(0) {
+    // initialize hilbert curve
+    std::vector<unsigned int> bin_to_idx(HILBERT_GRID_DIM * HILBERT_GRID_DIM * HILBERT_GRID_DIM);
+    for (int i = 0; i < HILBERT_GRID_DIM; i++) {
+        for (int j = 0; j < HILBERT_GRID_DIM; j++) {
+            for (int k = 0; k < HILBERT_GRID_DIM; k++) {
+
+                bitmask_t hilbert_coords[3];
+                hilbert_coords[0] = i;
+                hilbert_coords[1] = j;
+                hilbert_coords[2] = k;
+
+                unsigned int bin = static_cast<unsigned int>(hilbert_c2i(3, HILBERT_N_BITS, hilbert_coords));
+                bin_to_idx[i * HILBERT_GRID_DIM * HILBERT_GRID_DIM + j * HILBERT_GRID_DIM + k] = bin;
+            }
+        }
+    }
+
+    d_bin_to_idx_.copy_from(&bin_to_idx[0]);
```

Comment:
Not done in this PR, but we could make this buffer static across multiple instances of HilbertCurve to remove the N^3 memory usage per curve and investigate using larger grid sizes (reduced in #692)